Add sparse compressed meta tensor support

pearu opened the pull request:
As in the title.

Replaces https://github.com/pytorch/pytorch/pull/120498 and https://github.com/pytorch/pytorch/pull/120562 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120920
* __->__ #120707
* #120703



cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @jcaip

pearu committed:
Add sparse compressed meta tensor support

[ghstack-poisoned]

pearu committed:
Update on "Add sparse compressed meta tensor support"

Replaces https://github.com/pytorch/pytorch/pull/120498 and https://github.com/pytorch/pytorch/pull/120562 




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Add sparse compressed meta tensor support"

Replaces https://github.com/pytorch/pytorch/pull/120498 and https://github.com/pytorch/pytorch/pull/120562 




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Add sparse compressed meta tensor support"

Replaces https://github.com/pytorch/pytorch/pull/120498 and https://github.com/pytorch/pytorch/pull/120562 




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Add sparse compressed meta tensor support"

Replaces https://github.com/pytorch/pytorch/pull/120498 and https://github.com/pytorch/pytorch/pull/120562 




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Add sparse compressed meta tensor support"

Replaces https://github.com/pytorch/pytorch/pull/120498 and https://github.com/pytorch/pytorch/pull/120562 




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Add sparse compressed meta tensor support"

As in the title.

Replaces https://github.com/pytorch/pytorch/pull/120498 and https://github.com/pytorch/pytorch/pull/120562 




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

pearu committed:
Update on "Add sparse compressed meta tensor support"

As in the title.

Replaces https://github.com/pytorch/pytorch/pull/120498 and https://github.com/pytorch/pytorch/pull/120562 




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3040,12 +3041,17 @@ def hint(s):
+                if is_sparse_any(curr_t):
+                    for i, ss in enumerate(curr_t.size()):
+                        property_source = TensorPropertySource(src, TensorProperty.SIZE, i)
+                        track_symint(property_source, ss, constraint[i])
```
I feel there are some integers we have to track, like nnz?

pearu replied:
What is the exact function of tracking sparse tensor shape values and other integer parameters? Namely, a sparse tensor can be viewed as a tuple of indices and values tensors. While the sparse tensor shape and nnz is related to the shapes of indices and values, should we also track the strides of indices and values tensors (we do support non-contiguous values, for instance)?

pearu committed:
Update on "Add sparse compressed meta tensor support"

As in the title.

Replaces https://github.com/pytorch/pytorch/pull/120498 and https://github.com/pytorch/pytorch/pull/120562 




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer jcaip

[ghstack-poisoned]

ezyang reviewed (approved):
jolly good

ezyang commented:
@pytorchbot merge

laithsakka commented:
with this change shall i not graph break on    torch.sparse_compressed_tensor? in 
https://github.com/pytorch/pytorch/pull/120458/ 
@pearu

pearu commented:
> with this change shall i not graph break on torch.sparse_compressed_tensor? in #120458

If it is not necessary, then no. Note that adding sparse compressed tensors support to torch compiler is WIP, see e.g. https://github.com/pytorch/pytorch/pull/117907
